### PR TITLE
[Feature] Support EfficientFormer

### PR DESCRIPTION
**Pull request by Ezra-Yu**
## Motivation

Support EfficientFormer, [Arxiv](https://arxiv.org/abs/2206.01191) | [Github](https://github.com/snap-research/EfficientFormer)

## Use cases (Optional)
```
    Example:
        >>> from mmcls.models import EfficientFormer
        >>> import torch
        >>> inputs = torch.rand((1, 3, 224, 224))
        >>> # build EfficientFormer backbone for classification task
        >>> model = EfficientFormer(arch="l1")
        >>> model.eval()
        >>> level_outputs = model(inputs)
        >>> for level_out in level_outputs:
        ...     print(tuple(level_out.shape))
        (1, 448, 49)
        >>> # build EfficientFormer backbone for downstream task
        >>> EfficientFormer(
        >>>    arch="l3",
        >>>    out_indices=(0, 1, 2, 3),
        >>>    reshape_last_feat=True)
        >>> model.eval()
        >>> level_outputs = model(inputs)
        >>> for level_out in level_outputs:
        ...     print(tuple(level_out.shape))
```
## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/models/backbones/efficientformer.py [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+                 drop_rate=.0,
+                 drop_path_rate=0.,
+                 use_layer_scale=True,
+                 vit_num=1,
+                 has_downsamper=False):
+    """generate EfficientFormer blocks for a stage."""
+    blocks = []
+    if has_downsamper:
+        blocks.append(
+            PatchEmbed(
```
Comment (mzr1996) [inline mmcls/models/backbones/efficientformer.py:327]: Although the implementation of `PatchEmbed` is convolution, `PatchEmbed` is usually used in the stem and the `stride` and `patch_size` are usually the same.
Using a `PatchEmbed` module as the downsample layer is a little strange. May use sequential convolution and normalization layers.

File: mmcls/models/backbones/efficientformer.py [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+        frozen_stages (int): Stages to be frozen (stop grad and set eval mode).
+            -1 means not freezing any parameters. Defaults to -1.
+        act_cfg (dict): The config dict for activation between pointwise
+            convolution. Defaults to ``dict(type='GELU')``.
+        drop_rate (float): Dropout rate. Defaults to 0.
+        drop_path_rate (float): Stochastic depth rate. Defaults to 0.
+        use_layer_scale (bool): Whether to use use_layer_scale in MetaFormer
+             block. Defaults to True.
+        init_cfg (dict, optional): Initialization config dict.
+            Defaults to None.
```
Comment (mzr1996) [inline mmcls/models/backbones/efficientformer.py:398]: The order of the docstring doesn't match the arguments list.

File: mmcls/models/backbones/efficientformer.py [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+        for idx, block in enumerate(self.network):
+            if idx == len(self.network) - 1:
+                N, _, H, W = x.shape
+                if self.downsamples[idx]:
+                    H, W = H // 2, W // 2
+            x = block(x)
+            if idx in self.out_indices:
+                norm_layer = getattr(self, f'norm{idx}')
+
+                if idx == len(self.network) - 1 and x.dim() == 3:
```
Comment (mzr1996) [inline mmcls/models/backbones/efficientformer.py:584]: `x.dim() == 3`? why do we need this condition?

File: mmcls/models/backbones/efficientformer.py [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+                    # in the last stage, if `self.reshape_last_feat``
+                    # is True, using reshape to NCHW then GN.
+                    # if not, LN then from B N C permute to B C N
+                    if self.reshape_last_feat:
```
Comment (mzr1996) [inline mmcls/models/backbones/efficientformer.py:588]: Please add a detailed explanation of the usage of `reshape_last_feat` in the docstring.

File: mmcls/models/backbones/efficientformer.py [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+                if self.downsamples[idx]:
+                    H, W = H // 2, W // 2
+            x = block(x)
+            if idx in self.out_indices:
+                norm_layer = getattr(self, f'norm{idx}')
+
+                if idx == len(self.network) - 1 and x.dim() == 3:
+                    # in the last stage, if `self.reshape_last_feat``
+                    # is True, using reshape to NCHW then GN.
+                    # if not, LN then from B N C permute to B C N
```
Comment (mzr1996) [inline mmcls/models/backbones/efficientformer.py:587]:
```suggestion
                    # is True, reshape the features to `BCHW` before the final normalization.
                    # if False, do normalization directly and permute the features to `BCL`.
```

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: mmcls/models/backbones/efficientformer.py (hunk 1 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/efficientformer.py]: done.

File: mmcls/models/backbones/efficientformer.py (hunk 3 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/efficientformer.py:593]: in case the `vit-num` is 0 in `arch`.

File: mmcls/models/backbones/efficientformer.py (hunk 4 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/efficientformer.py:599]: done.

File: mmcls/models/backbones/efficientformer.py (hunk 2 above)
Comment (Ezra-Yu) [reply mmcls/models/backbones/efficientformer.py:407]: done.

**Review by mzr1996 — APPROVED** [review]
LGTM
